### DISCUSSION TITLE
Convert Markdown to plain text?

### DISCUSSION THREAD

**Discussion by zoul** [markdoc/markdoc#298] [Q&A] [answered]
Is there a simple way to use Markdoc to convert regular Markdown (with no custom tags) to plain text? I know I can write a custom renderer, but before I jump into that, maybe there‘s a simpler way? Currently I use `remark`:

```typescript
const stripMarkdown = (mdown: string) =>
  String(remark().use(strip).processSync(mdown));
```

…but I want to stick to Markdoc and don’t want to depend on both toolchains at once.

**Comment by zoul** [comment]
So far I have this, adapted from the built-in HTML renderer, turned out to be easier than I thought:

```typescript
function renderPlainText(node: RenderableTreeNodes): string {
  if (typeof node === "string" || typeof node === "number") {
    return String(node);
  }

  if (Array.isArray(node)) {
    return node.map(renderPlainText).join("");
  }

  if (node === null || typeof node !== "object" || !Tag.isTag(node)) {
    return "";
  }

  const { name, children = [] } = node;

  if (!name) {
    return renderPlainText(children);
  }

  return children.length ? renderPlainText(children) : "";
}
```

**Comment by rpaul-stripe** [comment, marked as answer]
You can also use the `.walk()` method on a document AST to iterate over every node and then output the content of just the text nodes. With this approach, you can avoid having to do the transform step. This is a quick-and-dirty implementation that prints the plain text of each inline block on a separate line:

```
function toPlainText(doc) {
  let output = '';
  for (let node of doc.walk()) {
    if (node.type === 'inline') output += '\n';
    if (node.type === 'text') output += node.attributes.content;
  }
  
  return output;
}
```

**Reply by zoul** [reply]
This is nice, thank you!
